core/commands/storage/challenge/dc_challenge.go

```diff
@@ -0,0 +1,178 @@
+package challenge
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/tron-us/go-btfs-common/crypto"
+	guardpb "github.com/tron-us/go-btfs-common/protos/guard"
+	"github.com/tron-us/go-btfs-common/utils/grpc"
+	"github.com/tron-us/go-common/v2/json"
+
+	cmds "github.com/TRON-US/go-btfs-cmds"
+	config "github.com/TRON-US/go-btfs-config"
+	"github.com/TRON-US/go-btfs/core/commands/cmdenv"
+	"github.com/TRON-US/go-btfs/core/commands/storage/helper"
+	"github.com/TRON-US/go-btfs/repo"
+
+	"github.com/cenkalti/backoff/v4"
+	logging "github.com/ipfs/go-log"
+)
+
+type Question struct {
+	FileHash     string
+	ShardHash    string
+	HostPid      string
+	ChunkIndex   int32
+	Nonce        string
+	ExpectAnswer string
+}
+
+const (
+	hostReqChallengePeriod = 2 * time.Second
+	hostChallengeTimeout   = 10 * time.Minute
+)
+
+var challengeHostBo = func() *backoff.ExponentialBackOff {
+	bo := backoff.NewExponentialBackOff()
+	bo.InitialInterval = 2 * time.Second
+	bo.MaxElapsedTime = hostChallengeTimeout
+	bo.Multiplier = 1
+	bo.MaxInterval = 2 * time.Second
+	return bo
+}()
+
+var log = logging.Logger("dc_challenge")
+
+func RequestChallenge(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment, cfg *config.Config) {
+	timer := time.NewTimer(hostReqChallengePeriod)
+	defer timer.Stop()
+	for ; true; <-timer.C {
+		result, err := retrieveChallengeQuestion(req, res, env, cfg)
+		if err != nil || result != nil {
+			timer.Reset(hostReqChallengePeriod)
+		}
+	}
+}
+
+func retrieveChallengeQuestion(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment, cfg *config.Config) (*guardpb.Result, error) {
+	n, err := cmdenv.GetNode(env)
+	peerId := n.Identity.Pretty()
+
+	challengeReq := &guardpb.ChallengeJobRequest{
+		NodePid:     peerId,
+		RequestTime: time.Now().UTC(),
+	}
+	sig, err := crypto.Sign(n.PrivateKey, challengeReq)
+	if err != nil {
+		fmt.Println("crypto.Sign err", err)
+		return nil, err
+	}
+	challengeReq.Signature = sig
+
+	ctx, _ := helper.NewGoContext(req.Context)
+	var challengeResp *guardpb.ChallengeJobResponse
+	err = grpc.GuardClient(cfg.Services.GuardDomain).WithContext(ctx, func(ctx context.Context,
+		client guardpb.GuardServiceClient) error {
+		challengeResp, err = client.RequestForChallengeJob(ctx, challengeReq)
+		if err != nil {
+			fmt.Printf("client.SubmitRepairContract err for peer id {%s}: %v\n", challengeReq.NodePid, err)
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		fmt.Println("grpc.GuardClient err", err)
+		return nil, err
+	}
+	if challengeResp == nil || challengeResp.PackageQuestionsCount == 0 {
+		fmt.Printf("failed to reuqest challenge job for peer id: {%s}\n", n.Identity.Pretty())
+		return nil, fmt.Errorf("failed to reuqest challenge job for peer id: {%s}", n.Identity.Pretty())
+	}
+
+	rawData, err := repo.DefaultS3Connection.RetrieveData(challengeResp.PackageUrl)
```

Comment:
Go-BTFS client cannot use s3 connection to retrieve the file, it should use the http connection to get the file.

Reply:
Update the code to use http request to get all the questions.